fairness/metrics.py

```diff
@@ -1,11 +1,206 @@
+    >>> print(diff)
+    3.9797979797979797
     """
     probs = np.linspace(0, 1, num=100)
     eqf1 = np.quantile(data1, probs)
     eqf2 = np.quantile(data2, probs)
     unfair_value = np.max(np.abs(eqf1-eqf2))
-    return unfair_value
\ No newline at end of file
+    return unfair_value
+
+
+def unfairness(y_predict, x_sa_test):
+    """
+    Compute the unfairness value for a given fair output and multiple sensitive attributes data contening several modalities.
+
+    Parameters:
+    y_predict (array-like): Predicted (fair or not) output data.
+    x_ssa_test (array-like): Sensitive attribute data.
+
+    Returns:
+    float: Unfairness value in the dataset.
+
+    Example:
+    >>> y_predict = np.array([5, 0, 6, 7, 9])
+    >>> x_sa_test = np.array([[1, 2, 1, 1, 2], [0, 1, 2, 1, 0]]).T
+    >>> unf = unfairness(y_predict, x_sa_test)
+    >>> print(unf)
+    6.0
+    """
+    new_list = []
+    if len(np.shape(x_sa_test)) == 1:
+        sens_val = list(set(x_sa_test))
+        data1 = y_predict
+        lst_unfairness = []
+        for mod in sens_val:
+            data2 = y_predict[x_sa_test == mod]
+            lst_unfairness.append(diff_quantile(data1, data2))
+        new_list.append(max(lst_unfairness))
+    else :
+        for sens in x_sa_test.T:
+            sens_val = list(set(sens))
+            data1 = y_predict
+            lst_unfairness = []
+            for mod in sens_val:
+                data2 = y_predict[sens == mod]
+                lst_unfairness.append(diff_quantile(data1, data2))
+            new_list.append(max(lst_unfairness))
+    return max(new_list)
+
+def unfairness_multi(y_fair_dict, x_sa_test):
+    """
+    Compute unfairness values for multiple fair output datasets and multiple sensitive attributes datasets.
+
+    Parameters:
+    y_fair_dict (dict): A dictionary where keys represent sensitive features and values are arrays
+            containing the fair predictions corresponding to each sensitive feature.
+            Each sensitive feature's fairness adjustment is performed sequentially,
+            ensuring that each feature is treated fairly relative to the previous ones.
+    x_sa_test (array-like): Sensitive attribute data.
+
+    Returns:
+    dict: A dictionary containing unfairness values for each level of fairness.
+
+    Example:
+    >>> y_fair_dict = {'Base model':np.array([19,39,65]), 'sens_var_1':np.array([22,40,50]), 'sens_var_2':np.array([28,39,42])}
+    >>> x_sa_test = np.array([['blue', 2], ['red', 9], ['green', 5]])
+    >>> unfs_dict = unfairness_multi(y_fair_dict, x_sa_test)
+    >>> print(unfs_dict)
+    {'sens_var_0': 46.0, 'sens_var_1': 28.0, 'sens_var_2': 14.0}
+    """
+    unfairness_dict = {}
+    for i, y_fair in enumerate(y_fair_dict.values()):
+        result = unfairness(y_fair, x_sa_test)
+        unfairness_dict[f'sens_var_{i}'] = result
+    return unfairness_dict
+
+
+def unfairness_multi_permutations(permut_y_fair_dict, all_combs_x_sa_test):
```

Comment:
regroup this and the previous function. For example, create a function that does one job, then use this function within the permutations:

ie:
```suggestion
def calculate_unfairness():
    ...

def some_fancy_permuation_name():
    ...
    calculate_unfairness()
    
```


This makes it a lot easier to read, control and stack

Reply:
I tried something, I think I should reorganized this in a folder 'permutation'